[DataLoader] Typing Enforcement for DataPipe at construct-time

ejguan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54773 [DataLoader] Typing Doc
* #54544 [DataLoader] Typing Enforcement for DataPipe at runtime
* **#54066 [DataLoader] Typing Enforcement for DataPipe at construct-time**
* #54020 [DataLoader] Typing Enforcement for DataPipe at Compile-time
* #54299 [DataLoader] Implement issubtype for type hints

## Feature
- Add a decorator `construct_time_validation` to validate each input datapipe according to the corresponding type hint.

Differential Revision: [D27327236](https://our.internmc.facebook.com/intern/diff/D27327236)

ejguan committed:
[DataLoader] Typing Enforcement for DataPipe at construct-time

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"




[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"




[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"




[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"




[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"




[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"




[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"




[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"




[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"




[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"




[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"




[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"




[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"


## Feature
- Add a decorator `construct_time_validation` to validate each input datapipe according to the corresponding type hint.

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"


## Feature
- Add a decorator `construct_time_validation` to validate each input datapipe according to the corresponding type hint.

Differential Revision: [D27327236](https://our.internmc.facebook.com/intern/diff/D27327236)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"


## Feature
- Add a decorator `construct_time_validation` to validate each input datapipe according to the corresponding type hint.

Differential Revision: [D27327236](https://our.internmc.facebook.com/intern/diff/D27327236)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"


## Feature
- Add a decorator `construct_time_validation` to validate each input datapipe according to the corresponding type hint.

Differential Revision: [D27327236](https://our.internmc.facebook.com/intern/diff/D27327236)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"


## Feature
- Add a decorator `construct_time_validation` to validate each input datapipe according to the corresponding type hint.

Differential Revision: [D27327236](https://our.internmc.facebook.com/intern/diff/D27327236)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"


## Feature
- Add a decorator `construct_time_validation` to validate each input datapipe according to the corresponding type hint.

Differential Revision: [D27327236](https://our.internmc.facebook.com/intern/diff/D27327236)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"


## Feature
- Add a decorator `construct_time_validation` to validate each input datapipe according to the corresponding type hint.

Differential Revision: [D27327236](https://our.internmc.facebook.com/intern/diff/D27327236)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"


## Feature
- Add a decorator `construct_time_validation` to validate each input datapipe according to the corresponding type hint.

Differential Revision: [D27327236](https://our.internmc.facebook.com/intern/diff/D27327236)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"


## Feature
- Add a decorator `construct_time_validation` to validate each input datapipe according to the corresponding type hint.

Differential Revision: [D27327236](https://our.internmc.facebook.com/intern/diff/D27327236)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"


## Feature
- Add a decorator `construct_time_validation` to validate each input datapipe according to the corresponding type hint.

Differential Revision: [D27327236](https://our.internmc.facebook.com/intern/diff/D27327236)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"


## Feature
- Add a decorator `construct_time_validation` to validate each input datapipe according to the corresponding type hint.

Differential Revision: [D27327236](https://our.internmc.facebook.com/intern/diff/D27327236)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"


## Feature
- Add a decorator `construct_time_validation` to validate each input datapipe according to the corresponding type hint.

Differential Revision: [D27327236](https://our.internmc.facebook.com/intern/diff/D27327236)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"


## Feature
- Add a decorator `construct_time_validation` to validate each input datapipe according to the corresponding type hint.

Differential Revision: [D27327236](https://our.internmc.facebook.com/intern/diff/D27327236)

[ghstack-poisoned]

VitalyFedyunin commented on test/test_datapipe.py:
```diff
@@ -697,6 +697,42 @@ def __iter__(self) -> Iterator:
         self.assertTrue(dp.type.param == int)
 
+    def test_construct_time(self):
+        from torch.utils.data import construct_time_validation
```
Please move import to the header

VitalyFedyunin commented on torch/utils/data/decorator.py:
```diff
@@ -94,3 +102,34 @@ def deterministic_wrapper_fn(self, *args, **kwargs) -> IterDataPipe:
+        for argument_name, value in bound.arguments.items():
+            if argument_name in hints and isinstance(hints[argument_name], _DataPipeMeta):
+                hint = hints[argument_name]
+                if not isinstance(value, IterDataPipe):
```
super nit: Need to add MapDataPipe support eventually.

ejguan replied:
I will add MapDataPipe support in the short future, as I need to move metaclass to `Dataset` to support typing system for MapDataPipe. And, it will take longer time to resolve all the internal breaks.

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"


## Feature
- Add a decorator `construct_time_validation` to validate each input datapipe according to the corresponding type hint.

Differential Revision: [D27327236](https://our.internmc.facebook.com/intern/diff/D27327236)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"


## Feature
- Add a decorator `construct_time_validation` to validate each input datapipe according to the corresponding type hint.

Differential Revision: [D27327236](https://our.internmc.facebook.com/intern/diff/D27327236)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"


## Feature
- Add a decorator `construct_time_validation` to validate each input datapipe according to the corresponding type hint.

Differential Revision: [D27327236](https://our.internmc.facebook.com/intern/diff/D27327236)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"


## Feature
- Add a decorator `construct_time_validation` to validate each input datapipe according to the corresponding type hint.

Differential Revision: [D27327236](https://our.internmc.facebook.com/intern/diff/D27327236)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at construct-time"


## Feature
- Add a decorator `construct_time_validation` to validate each input datapipe according to the corresponding type hint.

Differential Revision: [D27327236](https://our.internmc.facebook.com/intern/diff/D27327236)

[ghstack-poisoned]
